### PR TITLE
In memory text document

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [/] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have not broken the cheatsheet

### CODE REVIEW

File: packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.test.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      uri,
+      languageId,
+      "foo\nbar\r\nbaz",
+    );
+
+    assert.equal(document.lineCount, 3);
+    assert.equal(document.eol, "CRLF");
+  });
+
+  test("lineAt", () => {
```
Comment (pokey) [inline packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.test.ts:49]: Would be good to have some edge cases. At least these doc content examples:

- `""`
- `"aaa"`
- `"\naaa"`
- `"aaa\n"`
- `"\naaa\n"`
- `"\n"`
- `"\n\n"`
- `"aaa\n\naaa"`
- `"aaa\nbbb\r\nccc"`

You don't need to test all attributes of the lines for these; I think just range and range including newline? Worth doing the usual fixture thing with a TestCase interface containing input and expected output to make these more readable. Just short repr of the expected output ranges is fine I think

**Review by pokey — COMMENTED** [review]
Just a few quick thoughts; still need to review edit code

File: packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  get text(): string {
+    return this._text;
+  }
+
+  get range(): Range {
+    return new Range(this._lines[0].range.start, this._lines.at(-1)!.range.end);
+  }
+
+  setTextInternal(text: string): void {
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts:51]: This is called from performEdits. The name should just indicate that it's not for public use.

File: packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.test.ts
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.test.ts:1]: True. I made is quite late though I forgot somethings apparently :D

File: packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pokey) [reply packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts:51]: What about exposing `performEdits` as a method on the text document? If you want to keep the meat of that function in a separate file, you could rework the standalone `performEdits` to just return the new contents (and the changes that it already returns)

File: packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.test.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.test.ts:65]: Tests added

File: packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts:51]: I guess we could do that? Just goes a bit against the interface we have for text document and text editor, but in this case it's an implementation details I think that's fine.

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/ide/inMemoryTextDocument/InMemoryTextDocument.ts]: Ok this is now implemented

**Comment by pokey** [comment]
Made a bunch of tweaks and added a failing test. I personally think the test should pass, but we should just see how a VSCode editor handles that situation and do the same thing

**Comment by AndreasArvidsson** [comment]
Oka your new test is now passing

**Review by pokey — APPROVED** [review]

**Comment by pokey** [comment]
> Ok your new test is now passing. I've tested and vscode is doing exactly what your test is specifying.

Ha nice. Yeah we rely on that behaviour in our range updater
